5.sc_Morphology_Logistic_Regression/1.train_models/scripts/train_single_class.py

```diff
@@ -0,0 +1,283 @@
+    )
+]
+
+
+# In[9]:
+
+
+# at random downsample the DMSO treatment to match the number of wells in the LPS treatment
+seed = 0
+# get the number of wells in the LPS treatment
+trt_wells = training_data[
+    training_data["oneb_Metadata_Treatment_Dose_Inhibitor_Dose"] == treatment
+].shape[0]
+# get the number of wells in the DMSO treatment
+dmso_wells = training_data[
+    training_data["oneb_Metadata_Treatment_Dose_Inhibitor_Dose"] == control
+].shape[0]
+# downsample the DMSO treatment to match the number of wells in the LPS treatment
+dmso_holdout = training_data[
+    training_data["oneb_Metadata_Treatment_Dose_Inhibitor_Dose"] == control
+].sample(n=trt_wells, random_state=seed)
+# remove the downsampled DMSO wells from the data
+training_data = training_data.drop(dmso_holdout.index)
+
+
+# In[10]:
+
+
+# define metadata columns
+# subset each column that contains metadata
+metadata = training_data.filter(regex="Metadata")
+# drop all metadata columns
+data_x = training_data.drop(metadata.columns, axis=1)
+labeled_data = training_data["oneb_Metadata_Treatment_Dose_Inhibitor_Dose"]
+
+
+# In[11]:
+
+
+# specify model types, feature types, and phenotypic classes
+model_types = ["final", "shuffled_baseline"]
+feature_types = ["CP"]
+phenotypic_classes = [f"{treatment}"]
+# create stratified data sets forLPS k-fold cross validation
+straified_k_folds = StratifiedKFold(n_splits=3, shuffle=False)
+
+# create logistic regression model with following parameters
+log_reg_model = LogisticRegression(
+    penalty="elasticnet",
+    solver="saga",
+    max_iter=10,
+    n_jobs=-1,
+    random_state=0,
+    class_weight="balanced",
+)
+
+# specify parameters to tune for
+parameters = {"C": np.logspace(-3, 3, 7), "l1_ratio": np.linspace(0, 1, 11)}
+print(f"Parameters being tested during grid search: {parameters}\n")
+
+# create grid search with cross validation with hypertuning params
+grid_search_cv = GridSearchCV(
+    log_reg_model, parameters, cv=straified_k_folds, n_jobs=-1, scoring="f1_weighted"
+)
+
+# train model on training data on all combinations of model types, feature types, and phenotypic classes
+for model_type, feature_type, phenotypic_class in itertools.product(
+    model_types, feature_types, phenotypic_classes
+):
+    phenotypic_class_counts = training_data.loc[
+        training_data["oneb_Metadata_Treatment_Dose_Inhibitor_Dose"] == phenotypic_class
+    ].shape[0]
+    print(
+        f"Training {model_type} model on {feature_type} features for {phenotypic_class} with {phenotypic_class_counts} samples"
+    )
+
+    if model_type == "shuffled_baseline":
+        for column in data_x:
+            np.random.shuffle(data_x[column].values)
+
+    with parallel_backend("multiprocessing"):
+        with warnings.catch_warnings():
+            warnings.filterwarnings(
+                "ignore", category=ConvergenceWarning, module="sklearn"
+            )
+            grid_search_cv = grid_search_cv.fit(data_x, labeled_data)
```

Comment:
Great job catching warnings here! I struggled so hard to find this solution when I was training models with ` grid_search_cv`. Hopefully it didn't take you as long haha